### PR TITLE
Allow to send a list of files as argument

### PR DESCRIPTION
**Pull request by nvuillam**
Fixes https://github.com/CodeNarc/CodeNarc/issues/714

Allows to send a list of files are arguments, to avoid to browse for Groovy files when we already know them , in order to improve performances.

Built results have the same format than FilesystemSourceAnalyzer

image

### CODE REVIEW

**Comment by nvuillam** [comment]
@chrismair i think we're (finally) good :)

**Review by chrismair — CHANGES_REQUESTED** [review]
Nice work. Looks promising.

File: src/main/groovy/org/codenarc/CodeNarc.groovy [hunk 1]
```diff
@@ -225,8 +230,10 @@ Usage: java org.codenarc.CodeNarc [OPTIONS]
     }
 
     protected void setDefaultsIfNecessary() {
-        baseDir = baseDir ?: '.'
-        includes = includes  ?: '**/*.groovy'
```
[the remark this answers is not in the document]
Comment (nvuillam) [reply src/main/groovy/org/codenarc/CodeNarc.groovy:229]: done :)

File: src/main/groovy/org/codenarc/CodeNarc.groovy [hunk 2]
```diff
@@ -247,13 +254,26 @@ Usage: java org.codenarc.CodeNarc [OPTIONS]
      * @return a configured SourceAnalyzer instance
      */
     protected SourceAnalyzer createSourceAnalyzer() {
-        new FilesystemSourceAnalyzer(baseDirectory: baseDir, includes: includes, excludes: excludes, failOnError: failOnError)
```
[the remark this answers is not in the document]
Comment (nvuillam) [reply src/main/groovy/org/codenarc/CodeNarc.groovy:250]: done :)

File: src/main/groovy/org/codenarc/results/DirectoryResults.groovy [hunk 3]
```diff
@@ -62,6 +62,54 @@ class DirectoryResults implements Results {
         children.add(child)
     }
 
+    /**
+     * Add file results to current DirectoryResult instance, or recursively in a child Directory result
+     * Child DirectoryResults are created if not existing
+     */
```
[the remark this answers is not in the document]
Comment (nvuillam) [reply src/main/groovy/org/codenarc/results/DirectoryResults.groovy:69]: Done :)

File: src/main/groovy/org/codenarc/results/DirectoryResults.groovy [hunk 4]
```diff
@@ -62,6 +62,54 @@ class DirectoryResults implements Results {
...
+                // Now that sub directory results exists, call again the same method
+                this.addFileResultRecursive(fileRes)
+            }
+        }
+    }
+
+    /**
+     * Make sure that all necessary DirectoryResults for a given path are existing
+     * Ex: Create directory results for dir1 , then dir1/subdir1, then /dir1/subdir1/sub-subdir1 ...
+     */
```
[the remark this answers is not in the document]
Comment (nvuillam) [reply src/main/groovy/org/codenarc/results/DirectoryResults.groovy:98]: Set to private :)

File: src/test/groovy/org/codenarc/analyzer/FilesSourceAnalyzerTest.groovy [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        assert testCountRule.count == 5
+        assert results.getNumberOfFilesWithViolations(3) == 5
+        assert results.totalNumberOfFiles == 5
+
+        // Verify that the directory structure is properly reflected within the results
+        assert childResultsClasses(results) == [DirectoryResults]
+        def top = results.children[0]
+        assertEqualSets(childResultsClasses(top), [FileResults, DirectoryResults, DirectoryResults])
+    }
+
```
[the remark this answers is not in the document]
Comment (nvuillam) [reply src/test/groovy/org/codenarc/analyzer/FilesSourceAnalyzerTest.groovy:127]: Done :)

**Comment by nvuillam** [comment]
@chrismair it should be ok now :)

**Comment by chrismair** [comment]
Thanks for the contribution!

**Pull request merged**